### ISSUE TITLE
proposed ENH: object returned from compound Artist has `legend` attribute 

### ISSUE THREAD

**Issue by efiring** [matplotlib/matplotlib#8389]
Specifying Artists in the desired order for a legend requires isolating the Artists from whatever is returned by a given plotting function. (#8283)  It would be easier if the returned object could be passed directly.  This could be done if the returned object gained a `legend` attribute encapsulating the handles and corresponding labels.

For example, plot returns a list.  Instead, it could return a minimal list subclass with a `legend` attribute pointing to a minimal class which might have `legend.artists` and `legend.labels` tuples.  Then the legend function would look at each entry in its list of "handles", and if the entry contains a `legend` attribute, it would use it's tuples, appending their info to its own lists; otherwise, it would use the handle directly, as it does at present.

A more complex case is the infamous `hist`, returning a tuple with contents that differ depending on how the function was called.  Instead, it could return a minimal subclass of tuple with a `legend` attribute as described above. 

I think this approach could be implemented while maintaining backwards compatibility, and that it also provides a framework for improving the API in other ways.  For example, the object returned by `hist` could additionally be extended (subclass of namedtuple?) to allow access to its contents via names in addition to via indexing.

**Comment by QuLogic** [comment]
Instead of subclassing, you could combine this with the second proposal of #8283? Flatten the list and then ignore everything that didn't have a `legend` attribute.

(Though I do think moving the API toward lightweight `namedtuple` results does sound good.)

**Comment by phobson** [comment]
I was thinking along similar lines as @QuLogic. I really like the idea of returning `nametuples`

**Comment by story645** [comment]
@efiring and @QuLogic  can you provide a psuedocode like example of what you're proposing? I'm finding it a little hard to follow just from the description.

**Comment by tacaswell** [comment]
To some degree `plot` is always going to be an outlier here (thanks MATLAB), but I think we should beat everything else into returning a single (composite) artist and let the legend handler code sort it out.

One major hitch is that a tuple of handles means 'merge all of these into a single entry'.

**Comment by anntzer** [comment]
I have some slightly horrible proposal to do the "beat everything else into returning a single (composite) artist" with a proper (likely veeeery long) deprecation mechanism (actually it could just work as well for `plot` too).

Say we have a method `ax.foo(...)` which used to return a tuple, and we now want it to return a composite artist.  Fortunately, the tuple public API is relatively small -- it's just a bunch of dunders (which probably don't make sense for composite artists anyways), and `count` and `index` (which we can reasonably avoid as method names).  We simply implement these methods on the composite artist class, with a deprecation warning raised when they are called.
